### PR TITLE
Ajout projet mlExplore

### PR DESCRIPTION
**Pull request by GillesDeLille**
*Voici une modeste contribution pour notre communauté qui me motive tant.
En matière de communication, tout ou presque pour moi est une première. J'espère que la "pullRequest" est réalisée précisément comme vous le souhaitez et à bon port...*

Il s'agit d'un projet d'abord d'intérêt pédagogique, avec comme fil conducteur :

Proposer une aide aux utilisateurs pour choisir des modèles de Marchine Learning à partir de la nature des données dont ils disposent.

J'ai rédigé un README assez conséquent.

Cordialement,
Gilles Teixidor
gilles.teixidor@dgfip.finances.gouv.fr

### CODE REVIEW

**Comment by pnavaro** [comment]
Bonjour, tout d'abord je trouve le projet intéressant. J'ai juste une ou deux questions. Pourquoi ici proposez-vous de mélanger du R avec du Python ? Je trouve que cela complique beaucoup l'installation. De plus [Ranger](https://github.com/imbs-hl/ranger) sera beaucoup plus efficace que scikit-learn ici. Je pense que ce serait intéressant de séparer les deux approches R et Python plutot que de les mélanger. Il y a beaucoup d'alternative à Shiny en Python mais c'est difficile de choisir.
Bien à vous
Pierre

**Comment by GillesDeLille** [comment]
Bonjour,

Tout d'abord, merci pour toutes les remarques que vous m'apportez.

Mon projet est avant tout pédagogique. Ce sera le fil conducteur de ma réponse.

Pourquoi mélanger R et Python :

- Avantages

- [ ] Réunir les développeurs dans un même projet. Pour le mener à bien, chacun va s'intéresser *vraiment* à la communauté *d'en face*
- [ ] **scikitlearn**. J'ai compris qu'en matière de ML, Python est populaire en grande partie grâce à l'apport de cette bibliothèque. Sa force pédagogique doit nous servir.
- [ ] De manière générale, on va trouver de bonnes (et mauvaises) implémentations dans les 2 langages. Je souhaite casser la paroi de verre entre les 2
- [ ] Une bonne part des lecteurs/contributeurs de Spyrales veut faire du ML. Pas frocément du R ni du Python. Beaucoup aussi trouve très bien les solutions clés en main de Google*. Enfin je crois, le sondage le dira...


- Inconvénients

- [ ] Install un peu + compliquée. C'est vrai. Un si mauvais moment à passer ?
- [ ] Efficacité du point de vue de l'utilisateur diminuée. C'est certain, si dans cette optique, je choisi mal les bibliothèques pour appliquer les modèles.

Je vais m'empresser** de partager votre conseil relatif à l'utilisation de **ranger**. D'ailleurs cette implémentation invite à travailler sous R. Shiny est alors parfait, non ;-) ?
- [ ] L'efficacité du point de vue du développeur est peut-être diminuée, mais je n'en suis pas sûr.

---
Notes :

_* J'en fait partie, mais j'ai pas les sous. Plus sérieusement, j'aimerais disposer un jour de nos propres modèles préentraînés._

_** Sur l'efficacité_
_Dans la section "Présentation des modèles", qui est encore vide actuellement, apparaîtra (du moins c'est mon souhait), en plus d'une description du modèle de ML :_

- _Les remarques, en termes d'efficacité notament, sur les différentes implémentations possibles_
- _Le temps d'exécution avec l'implémentation choisie (ce temps apparaît avec l'install option 2 dans la console de Rstudio)_

**Comment by pnavaro** [comment]
Vous avez raison Shiny est excellent mais il y a la même chose en Python comme Voilà https://voila-gallery.org. 

J'ai essayé de lancer l'application sans Rstudio avec la procédure suivante

J'ai utilisé le fichier conda téléchargeable ici https://github.com/pnavaro/la-piscine/blob/master/mlExplore/environment.yml

```
conda env create -f environment.yml
conda activate ml-explore
R -e "shiny::runApp('mlExplore')"
```
> Malheureusement j'ai un souci, les fichiers résultats n'apparaissent pas dans l'application, elles sont sauvés sous forme de fichiers png.

**Correction** : Cliquer sur le bouton `Modele`. Ça marche.

**Comment by GillesDeLille** [comment]
Pierre,
"Les fichiers résultats", s'il s'agit de la courbe de gain cumulée, la sauvegarde de l'image au format png est le comportement que j'attends (script randomForest.py).
Après si dans votre environnement conda le renderImage (output$imageGain) ne renvoie rien, j'ai peur de ne pas trop savoir quoi vous répondre.
Le png arrive-t-il bien dans le dossier **figures** ?
Que renvoie, dans votre environnement, la commande :

**normalizePath(file.path('./figures')**

**Comment by pnavaro** [comment]
C'est bon ca marche, j'ai le graphe. Il apparait dans "Modele". Désolé.

**Comment by GillesDeLille** [comment]
Pas de soucis, j'étais dessus pour intégrer l'implémentation **ranger** (qui va me plaire)

A ce propos, je ne vais pas prendre le temps de regarder dans l'immédiat tous les détails de son utilisation.
J'organise en premier lieu l'acceuil le plus naturel possible dans le projet de toute implementation nouvelle. Charge aux specialistes volontaires de celles-ci d'apporter les précisions nécessaires

L'algo tourne bien. Pour l'instant je bute sur l'affichage des résultats (c'est ballo...)

**Comment by Welsh44** [comment]
Bonjour @pnavaro et @GillesDeLille 
Je préviens je n'y connais rien en python et ML, ou pas grand chose.
@strainel m'a demandé une revue, ce que j'ai tenté de faire comme suit :
- j'ai récupéré le dossier mlExplore (qui me semble une très belle initiative)
- j'ai lancé le projet sous RStudio 1.2.5033 (R 3.6.2).

Bien évidemment, il manquait des paquets R : reticulate, shinyEffects et tictoc. installation faite, on y retourne.

Là dans la console on demande l'installation de Miniconda. Ce que j'ai accepté. On relance.

KO car pas moyen d'accéder au repository. Donc dansglobal.R j'ai ajouté une ligne `Sys.setenv(http_proxy = "xxxx")` qui correspond à mon environnement professionnel.

On y retourne et désormais l'installation des packages python se lance.... Ne pas couper l'application shiny sinon on recommence (je sais c'est ce que j'ai fait...).

Désormais, tout se lance mais il manque a priori deux paqets python : `pandas`et `sklearn` qu'il faut installer manuellement.

Bilan de la revue :
 - concept très intéressant
- pré-requis à préciser

**Review by Welsh44 — CHANGES_REQUESTED** [review]
Bonjour @pnavaro et @GillesDeLille 
Je préviens je n'y connais rien en python et ML, ou pas grand chose.
@strainel m'a demandé une revue, ce que j'ai tenté de faire comme suit :
- j'ai récupéré le dossier mlExplore (qui me semble une très belle initiative)
- j'ai lancé le projet sous RStudio 1.2.5033 (R 3.6.2).

Bien évidemment, il manquait des paquets R : reticulate, shinyEffects et tictoc. installation faite, on y retourne.

Là dans la console on demande l'installation de Miniconda. Ce que j'ai accepté. On relance.

KO car pas moyen d'accéder au repository. Donc dans `global.R` j'ai ajouté une ligne `Sys.setenv(http_proxy = "xxxx")` qui correspond à mon environnement professionnel.

On y retourne et désormais l'installation des packages python se lance.... Ne pas couper l'application shiny sinon on recommence (je sais c'est ce que j'ai fait...).

Désormais, tout se lance mais il manque a priori deux paquets python : `pandas`et `sklearn` qu'il faut installer manuellement.

Bilan de la revue :
 - concept très intéressant
- pré-requis à préciser

**Comment by GillesDeLille** [comment]
Bonjour Frédéric, (c'est bien vous Welsh44 ?)

Pour les prérequis, je prends au moins la partie la plus facile.
Le http_proxy, c'est pas ma tasse de thé. J'avais vu passer sur Tchap la contribution de guionnet@insa-toulouse.fr sur Renviron.site.
Peut être faudrait-il des conseils de paramétrage dans un cadre plus général sur Spyrales Web ?

Pour mon petit soucis d'affichage des résultats de ranger, je vais faire le minimum : sortir **prediction.error**. Pour les autres indicateurs on verra plus tard.

En résumé :
- [ ] Lister les librairies nécessaires (y compris celles pour faire joli...)
- [ ] sortie minimale des résultats

Content que le concept vous plaise,
Gilles

**Comment by Welsh44** [comment]
> 
> Bonjour Frédéric, (c'est bien vous Welsh44 ?)

Oui c'est bien moi :-)

> Pour les prérequis, je prends au moins la partie la plus facile.
> Le http_proxy, c'est pas ma tasse de thé. J'avais vu passer sur Tchap la contribution de [guionnet@insa-toulouse.fr](mailto:guionnet@insa-toulouse.fr) sur Renviron.site.

Entièrement d'accord. Mais là il fallait cette ligne pour installer Miniconda et les paquets. Juste à signalrr.

> Peut être faudrait-il des conseils de paramétrage dans un cadre plus général sur Spyrales Web ?

Bonne idée.

> Pour mon petit soucis d'affichage des résultats de ranger, je vais faire le minimum : sortir **prediction.error**. Pour les autres indicateurs on verra plus tard.
> 
> En résumé :
> 
>     * [ ]  Lister les librairies nécessaires (y compris celles pour faire joli...)
> 
>     * [ ]  sortie minimale des résultats
> 


> Content que le concept vous plaise,
> Gilles

**Comment by strainel** [comment]
PR fermée suite à une autre proposition prenant en compte ces échanges
